File: runtime/Java/src/org/antlr/v4/runtime/misc/InterpreterDataReader.java
/*
 * Copyright (c) 2012-2017 The ANTLR Project. All rights reserved.
 * Use of this file is governed by the BSD 3-clause license that
 * can be found in the LICENSE.txt file in the project root.
 */

package org.antlr.v4.runtime.misc;

import org.antlr.v4.runtime.Vocabulary;
import org.antlr.v4.runtime.VocabularyImpl;
import org.antlr.v4.runtime.atn.ATN;
import org.antlr.v4.runtime.atn.ATNDeserializer;
import org.antlr.v4.runtime.dfa.DFA;

import java.util.Collection;
import java.util.List;
import java.util.ArrayList;
import java.io.BufferedReader;
import java.io.FileReader;

// A class to read plain text interpreter data produced by ANTLR.
public class InterpreterDataReader {
	
	public static class InterpreterData {
	  ATN atn;
	  Vocabulary vocabulary;
	  List<String> ruleNames;
	  List<String> channels; // Only valid for lexer grammars.
	  List<String> modes; // ditto
	};
	
	/**
	 * The structure of the data file is very simple. Everything is line based with empty lines
	 * separating the different parts. For lexers the layout is:
	 * token literal names:
	 * ...
	 * 
	 * token symbolic names:
	 * ...
	 * 
	 * rule names:
	 * ...
	 * 
	 * channel names:
	 * ...
	 * 
	 * mode names:
	 * ...
	 * 
	 * atn:
	 * <a single line with comma separated int values> enclosed in a pair of squared brackets.
	 * 
	 * Data for a parser does not contain channel and mode names.
	 */
	public static InterpreterData parseFile(String fileName) {
		InterpreterData result = new InterpreterData();
		result.ruleNames = new ArrayList<String>();
		
		try (BufferedReader br = new BufferedReader(new FileReader(fileName))) {
		    String line;
		  	List<String> literalNames = new ArrayList<String>();
		  	List<String> symbolicNames = new ArrayList<String>();
		
			line = br.readLine();
			if ( !line.equals("token literal names:") )
				throw new RuntimeException("Unexpected data entry");
		    while ((line = br.readLine()) != null) {
		       if ( line.isEmpty() )
					break;
				literalNames.add(line.equals("null") ? "" : line);
		    }
		
			line = br.readLine();
			if ( !line.equals("token symbolic names:") )
				throw new RuntimeException("Unexpected data entry");
		    while ((line = br.readLine()) != null) {
		       if ( line.isEmpty() )
					break;
				symbolicNames.add(line.equals("null") ? "" : line);
		    }

		  	result.vocabulary = new VocabularyImpl(literalNames.toArray(new String[0]), symbolicNames.toArray(new String[0]));

			line = br.readLine();
			if ( !line.equals("rule names:") )
				throw new RuntimeException("Unexpected data entry");
		    while ((line = br.readLine()) != null) {
		       if ( line.isEmpty() )
					break;
				result.ruleNames.add(line);
		    }
		    
			if ( line.equals("channel names:") ) { // Additional lexer data.
				result.channels = new ArrayList<String>();
			    while ((line = br.readLine()) != null) {
			       if ( line.isEmpty() )
						break;
					result.channels.add(line);
			    }

				line = br.readLine();
				if ( !line.equals("mode names:") )
					throw new RuntimeException("Unexpected data entry");
				result.modes = new ArrayList<String>();
			    while ((line = br.readLine()) != null) {
			       if ( line.isEmpty() )
						break;
					result.modes.add(line);
			    }
			}

		  	line = br.readLine();
		  	if ( !line.equals("atn:") )
		  		throw new RuntimeException("Unexpected data entry");
			line = br.readLine();
			String[] elements = line.split(",");
	  		char[] serializedATN = new char[elements.length];

			for (int i = 0; i < elements.length; ++i) {
				int value;
				String element = elements[i];
				if ( element.startsWith("[") )
					value = Integer.parseInt(element.substring(1).trim());
				else if ( element.endsWith("]") )
					value = Integer.parseInt(element.substring(0, element.length() - 1).trim());
				else
					value = Integer.parseInt(element.trim());
				serializedATN[i] = (char)value;					
			}

		  	ATNDeserializer deserializer = new ATNDeserializer();
		  	result.atn = deserializer.deserialize(serializedATN);
		}
		catch (java.io.IOException e) {
			// We just swallow the error and return empty objects instead.
		}
		
		return result;
	}
	
}

File: runtime/Cpp/runtime/src/misc/InterpreterDataReader.cpp
/* Copyright (c) 2012-2017 The ANTLR Project. All rights reserved.
 * Use of this file is governed by the BSD 3-clause license that
 * can be found in the LICENSE.txt file in the project root.
 */

#include "atn/ATN.h"
#include "atn/ATNDeserializer.h"
#include "Vocabulary.h"
#include "VocabularyImpl.h"

#include "misc/InterpreterDataReader.h"

using namespace antlr4::dfa;
using namespace antlr4::atn;
using namespace antlr4::misc;
using namespace antlr4;

namespace {

  class InterpreterDataVocabulary final : public Vocabulary {
  public:
    InterpreterDataVocabulary(std::vector<std::string> literalNames, std::vector<std::string> symbolicNames)
        : Vocabulary(std::max(literalNames.size(), symbolicNames.size()) - 1),
          literalNames(std::move(literalNames)), symbolicNames(std::move(symbolicNames)),
          literalNamesView(this->literalNames.size()), symbolicNamesView(this->symbolicNames.size()),
          _impl(antlrcpp::Span<const std::string_view>(literalNamesView.data(), literalNamesView.size()), antlrcpp::Span<const std::string_view>(symbolicNamesView.data(), symbolicNamesView.size())) {
      for (size_t index = 0; index < this->literalNames.size(); index++) {
        literalNamesView[index] = this->literalNames[index];
      }
      for (size_t index = 0; index < this->symbolicNames.size(); index++) {
        symbolicNamesView[index] = this->symbolicNames[index];
      }
    }

    std::string_view getLiteralName(size_t tokenType) const override {
      return _impl.getLiteralName(tokenType);
    }

    std::string_view getSymbolicName(size_t tokenType) const override {
      return _impl.getSymbolicName(tokenType);
    }

    std::string getDisplayName(size_t tokenType) const override {
      return _impl.getDisplayName(tokenType);
    }

  private:
    std::vector<std::string> literalNames;
    std::vector<std::string> symbolicNames;
    std::vector<std::string_view> literalNamesView;
    std::vector<std::string_view> symbolicNamesView;
    VocabularyImpl _impl;
  };

}

InterpreterData InterpreterDataReader::parseFile(std::string const& fileName) {
  // The structure of the data file is very simple. Everything is line based with empty lines
  // separating the different parts. For lexers the layout is:
  // token literal names:
  // ...
  //
  // token symbolic names:
  // ...
  //
  // rule names:
  // ...
  //
  // channel names:
  // ...
  //
  // mode names:
  // ...
  //
  // atn:
  // <a single line with comma separated int values> enclosed in a pair of squared brackets.
  //
  // Data for a parser does not contain channel and mode names.

  std::ifstream input(fileName);
  if (!input.good())
    return {};

  std::vector<std::string> literalNames;
  std::vector<std::string> symbolicNames;

  std::string line;

  std::getline(input, line, '\n');
  assert(line == "token literal names:");
  while (true) {
    std::getline(input, line, '\n');
    if (line.empty())
      break;

    literalNames.push_back(line == "null" ? "" : line);
  };

  std::getline(input, line, '\n');
  assert(line == "token symbolic names:");
  while (true) {
    std::getline(input, line, '\n');
    if (line.empty())
      break;

    symbolicNames.push_back(line == "null" ? "" : line);
  };
  InterpreterData result;
  result.vocabulary = std::make_unique<InterpreterDataVocabulary>(std::move(literalNames), std::move(symbolicNames));

  std::getline(input, line, '\n');
  assert(line == "rule names:");
  while (true) {
    std::getline(input, line, '\n');
    if (line.empty())
      break;

    result.ruleNames.push_back(line);
  };

  std::getline(input, line, '\n');
  if (line == "channel names:") {
    while (true) {
      std::getline(input, line, '\n');
      if (line.empty())
        break;

      result.channels.push_back(line);
    };

    std::getline(input, line, '\n');
    assert(line == "mode names:");
    while (true) {
      std::getline(input, line, '\n');
      if (line.empty())
        break;

      result.modes.push_back(line);
    };
  }

  std::getline(input, line, '\n');
  assert(line == "atn:");
  std::getline(input, line, '\n');
  std::stringstream tokenizer(line);
  std::string value;
  while (tokenizer.good()) {
    std::getline(tokenizer, value, ',');
    unsigned long number;
    if (value[0] == '[')
      number = std::strtoul(&value[1], nullptr, 10);
    else
      number = std::strtoul(value.c_str(), nullptr, 10);
    result.serializedATN.push_back(static_cast<int32_t>(number));
  }


  ATNDeserializer deserializer;
  result.atn = deserializer.deserialize(antlrcpp::Span<const int32_t>(result.serializedATN.data(), result.serializedATN.size()));
  return result;
}
